cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,409 @@ struct ReplaceWithMaskFunctor {
     }
     return ReplaceWithMask<Type>::ExecArrayMask(ctx, array, mask, replacements, output);
   }
+
+  static std::shared_ptr<KernelSignature> GetSignature(detail::GetTypeId get_id){
+    return KernelSignature::Make(
+        {InputType::Array(get_id.id), InputType(boolean()), InputType(get_id.id)},
+        OutputType(FirstType));
+  }
 };
 
-}  // namespace
+// This is for fixed-size types only
+template <typename Type>
+void FillNullInDirectionImpl(const ArrayData current_chunk, const uint8_t* null_bitmap,
+                             ArrayData* output, int8_t direction,
+                             ArrayData last_valid_value_chunk,
+                             int64_t* last_valid_value_offset) {
+  uint8_t* out_bitmap = output->buffers[0]->mutable_data();
+  uint8_t* out_values = output->buffers[1]->mutable_data();
+  arrow::internal::CopyBitmap(current_chunk.buffers[0]->data(), current_chunk.offset,
+                              current_chunk.length, out_bitmap, output->offset);
+  ReplaceWithMask<Type>::CopyData(*current_chunk.type, out_values,
+                                  /*out_offset=*/output->offset, current_chunk,
+                                  /*in_offset=*/0, current_chunk.length);
 
-const FunctionDoc replace_with_mask_doc(
-    "Replace items selected with a mask",
-    ("Given an array and a boolean mask (either scalar or of equal length),\n"
-     "along with replacement values (either scalar or array),\n"
-     "each element of the array for which the corresponding mask element is\n"
-     "true will be replaced by the next value from the replacements,\n"
-     "or with null if the mask is null.\n"
-     "Hence, for replacement arrays, len(replacements) == sum(mask == true)."),
-    {"values", "mask", "replacements"});
+  bool has_fill_value = *last_valid_value_offset != -1;
+  int64_t write_offset = direction == 1 ? 0 : current_chunk.length - 1;
+  int64_t bitmap_offset = 0;
 
-void RegisterVectorReplace(FunctionRegistry* registry) {
-  auto func = std::make_shared<VectorFunction>("replace_with_mask", Arity::Ternary(),
-                                               &replace_with_mask_doc);
+  arrow::internal::OptionalBitBlockCounter counter(null_bitmap, output->offset,
+                                                   current_chunk.length);
+
+  while (bitmap_offset < current_chunk.length) {
+    BitBlockCount block = counter.NextBlock();
+    if (block.AllSet()) {
+      *last_valid_value_offset =
+          write_offset + direction * (block.length - 1 + bitmap_offset);
+      has_fill_value = true;
+      last_valid_value_chunk = current_chunk;
```

Comment:
We should declare something like this above:

```cpp
const ArrayData* source_chunk = &last_valid_value_chunk;
```

then here we can just assign to the pointer.

All together, something like this will let you avoid copying the ArrayData around (which is still somewhat expensive since it contains shared_ptrs):

```diff
diff --git a/cpp/src/arrow/compute/kernels/vector_replace.cc b/cpp/src/arrow/compute/kernels/vector_replace.cc
index 6651e4859..92846fc25 100644
--- a/cpp/src/arrow/compute/kernels/vector_replace.cc
+++ b/cpp/src/arrow/compute/kernels/vector_replace.cc
@@ -452,9 +452,9 @@ struct ReplaceWithMaskFunctor {
 
 // This is for fixed-size types only
 template <typename Type>
-void FillNullInDirectionImpl(const ArrayData current_chunk, const uint8_t* null_bitmap,
+void FillNullInDirectionImpl(const ArrayData& current_chunk, const uint8_t* null_bitmap,
                              ArrayData* output, int8_t direction,
-                             ArrayData last_valid_value_chunk,
+                             const ArrayData& last_valid_value_chunk,
                              int64_t* last_valid_value_offset) {
   uint8_t* out_bitmap = output->buffers[0]->mutable_data();
   uint8_t* out_values = output->buffers[1]->mutable_data();
@@ -464,6 +464,7 @@ void FillNullInDirectionImpl(const ArrayData current_chunk, const uint8_t* null_
                                   /*out_offset=*/output->offset, current_chunk,
                                   /*in_offset=*/0, current_chunk.length);
 
+  const ArrayData* source_chunk = &last_valid_value_chunk;
   bool has_fill_value = *last_valid_value_offset != -1;
   int64_t write_offset = direction == 1 ? 0 : current_chunk.length - 1;
   int64_t bitmap_offset = 0;
@@ -477,7 +478,7 @@ void FillNullInDirectionImpl(const ArrayData current_chunk, const uint8_t* null_
       *last_valid_value_offset =
           write_offset + direction * (block.length - 1 + bitmap_offset);
       has_fill_value = true;
-      last_valid_value_chunk = current_chunk;
+      source_chunk = &current_chunk;
     } else {
       uint64_t block_start_offset = write_offset + direction * bitmap_offset;
       uint64_t write_value_offset = block_start_offset;
@@ -485,8 +486,8 @@ void FillNullInDirectionImpl(const ArrayData current_chunk, const uint8_t* null_
         auto current_bit = bit_util::GetBit(null_bitmap, bitmap_offset + i);
         if (!current_bit) {
           if (has_fill_value) {
-            ReplaceWithMask<Type>::CopyData(*current_chunk.type, out_values,
-                                            write_value_offset, last_valid_value_chunk,
+            ReplaceWithMask<Type>::CopyData(*source_chunk->type, out_values,
+                                            write_value_offset, *source_chunk,
                                             *last_valid_value_offset,
                                             /*length=*/1);
             bit_util::SetBitTo(out_bitmap, write_value_offset, true);
@@ -494,7 +495,7 @@ void FillNullInDirectionImpl(const ArrayData current_chunk, const uint8_t* null_
         } else {
           has_fill_value = true;
           *last_valid_value_offset = write_value_offset;
-          last_valid_value_chunk = current_chunk;
+          source_chunk = &current_chunk;
         }
       }
     }
```